sdks/python/apache_beam/runners/direct/direct_userstate.py

```diff
@@ -0,0 +1,115 @@
+    self.step_context = step_context
+    self.dofn = dofn
+    self.key_coder = key_coder
+
+    self.all_state_specs, self.all_timer_specs = (
+        UserStateUtils.get_dofn_specs(dofn))
+    self.state_tags = {}
+    for state_spec in self.all_state_specs:
+      state_key = 'user/%s' % state_spec.name
+      if isinstance(state_spec, BagStateSpec):
+        state_tag = _ListStateTag(state_key)
+      elif isinstance(state_spec, CombiningValueStateSpec):
+        state_tag = _ListStateTag(state_key)
+      else:
+        raise ValueError('Invalid state spec: %s' % state_spec)
+      self.state_tags[state_spec] = state_tag
+
+    self.cached_states = {}
+    self.cached_timers = {}
+
+  def get_timer(self, timer_spec, key, window):
+    assert timer_spec in self.all_timer_specs
+    encoded_key = self.key_coder.encode(key)
+    cache_key = (encoded_key, window, timer_spec)
+    if cache_key not in self.cached_timers:
+      self.cached_timers[cache_key] = RuntimeTimer(timer_spec)
+    return self.cached_timers[cache_key]
+
+  def get_state(self, state_spec, key, window):
+    assert state_spec in self.all_state_specs
+    encoded_key = self.key_coder.encode(key)
+    cache_key = (encoded_key, window, state_spec)
+    if cache_key not in self.cached_states:
+      state_tag = self.state_tags[state_spec]
+      value_accessor = (
+          lambda: self._get_underlying_state(state_spec, key, window))
+      self.cached_states[cache_key] = (
+          RuntimeState.for_spec(state_spec, state_tag, value_accessor))
+    return self.cached_states[cache_key]
+
+  def _get_underlying_state(self, state_spec, key, window):
+    state_tag = self.state_tags[state_spec]
+    encoded_key = self.key_coder.encode(key)
+    return (self.step_context.get_keyed_state(encoded_key)
+            .get_state(window, state_tag))
+
+  def commit(self):
+    # Commit state modifications.
+    for cache_key, runtime_state in self.cached_states.items():
+      encoded_key, window, state_spec = cache_key
+      state = self.step_context.get_keyed_state(encoded_key)
+      state_tag = self.state_tags[state_spec]
+      if isinstance(state_spec, BagStateSpec):
+        if runtime_state._cleared:
+          state.clear_state(window, state_tag)
+        for new_value in runtime_state._new_values:
+          state.add_state(window, state_tag, new_value)
+      elif isinstance(state_spec, CombiningValueStateSpec):
+        if runtime_state._modified:
+          state.clear_state(window, state_tag)
+          state.add_state(
+              window, state_tag,
+              state_spec.coder.encode(runtime_state._current_accumulator))
+      else:
+        raise ValueError('Invalid state spec: %s' % state_spec)
+
+    # Commit new timers.
+    for cache_key, runtime_timer in self.cached_timers.items():
+      encoded_key, window, timer_spec = cache_key
+      state = self.step_context.get_keyed_state(encoded_key)
+      timer_name = 'user/%s' % timer_spec.name
```

Comment:
This is because state is accessed through state tag objects, and timers are accessed through string keys.  The logic for creating particular state tag objects has a lot of special casing, so we create the tags once during init so that we can use it everywhere else.  There is no value in caching the string key for the timers until we need to use it, which is currently only on commit.  It is unfortunately a little inconsistent, but I don't see a cleaner way to do this (there is no value in moving this string concatenation to `__init__`, and `__init__` is still the cleanest place to do the state tag construction).  Let me know if you see an alternative.

Reply:
OK, thanks for the explanation.